src/modules/expression/binop/mod.rs

```diff
@@ -21,39 +22,47 @@ pub trait BinOp: SyntaxModule<ParserMetadata> {
-}
-
-#[macro_export]
-macro_rules! handle_binop {
-    (@internal type: Array) => {
-        Type::Array(_)
-    };
-
-    (@internal type: $type:ident) => {
-        Type::$type
-    };
 
-    ($meta:expr, $op_name:expr, $left:expr, $right:expr, [$($type_match:ident),+]) => {{
-        let left_match = matches!($left.get_type(), $(handle_binop!(@internal type: $type_match))|*);
-        let right_match = matches!($right.get_type(), $(handle_binop!(@internal type: $type_match))|*);
-        if !left_match || !right_match || $left.get_type() != $right.get_type() {
-            let pos = $crate::modules::expression::binop::get_binop_position_info($meta, &$left, &$right);
-            let message = Message::new_err_at_position($meta, pos);
-            error_type_match!($meta, message, $op_name, $left, $right, [$($type_match),+])
+    fn typecheck_allowed_types(
+        meta: &mut ParserMetadata,
+        operator: &str,
+        left: &Expr,
+        right: &Expr,
+        allowed_types: &[Type],
+    ) -> Result<Type, Failure> {
+        let left_type = left.get_type();
+        let right_type = right.get_type();
+        let left_match = allowed_types.iter().any(|types| left_type.is_allowed_in(types));
+        let right_match = allowed_types.iter().any(|types| right_type.is_allowed_in(types));
+        if !left_match || !right_match {
+            let pos = get_binop_position_info(meta, left, right);
+            let message = Message::new_err_at_position(meta, pos);
+            let msg = format!("Cannot perform {operator} on value of type '{left_type}' and value of type '{right_type}'");
+            let pretty_types = Type::pretty_join(allowed_types, "and");
+            let sentence = if allowed_types.len() == 1 { "Allowed type is" } else { "Allowed types are" };
+            let comment = format!("{sentence} {pretty_types}.");
+            Err(Failure::Loud(message.message(msg).comment(comment)))
         } else {
-            Ok($left.get_type())
+            Self::typecheck_equality(meta, left, right)
         }
-    }};
+    }
 
-    ($meta:expr, $op_name:expr, $left:expr, $right:expr) => {{
-        if $left.get_type() != $right.get_type() {
-            let pos = $crate::modules::expression::binop::get_binop_position_info($meta, &$left, &$right);
-            let message = Message::new_err_at_position($meta, pos);
-            error_type_match!($meta, message, $op_name, $left, $right)
+    fn typecheck_equality(
+        meta: &mut ParserMetadata,
+        left: &Expr,
+        right: &Expr,
+    ) -> Result<Type, Failure> {
+        let left_type = left.get_type();
+        let right_type = right.get_type();
+        if left_type != right_type {
+            let pos = get_binop_position_info(meta, left, right);
+            let message = Message::new_err_at_position(meta, pos);
+            let msg = format!("Expected both operands to be of the same type, but got '{left_type}' and '{right_type}'.");
```

Comment:
I don't like that we use variable named "message" twice but the second time in a shorthand version. Why not pass this formated string directly into a message object?

Reply:
The `message` variable is the structure that contains a message that is to be displayed to the user. The `msg` is a variable containing the message contents. It could be called `message_content` or something similar.